javatests/arcs/core/analysis/ExpressionClaimDeducerTest.kt

```diff
@@ -0,0 +1,370 @@
+          )
+      )
+    )
+  }
+
+  @Test
+  fun field_access_nested_expr() {
+    val expr = PaxelParser.parse("(new Foo {x: foo, y: bar}).x")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Equal("foo"))
+  }
+
+  @Test
+  fun field_access_nested_multi_field_expr() {
+    val expr = PaxelParser.parse("(new Foo {a: foo.x + foo.y, b: foo.y + foo.z}).a")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        listOf("foo", "x"),
+        listOf("foo", "y")
+      )
+    )
+  }
+
+  @Test
+  fun new_nested() {
+    val expr = PaxelParser.parse(
+      """
+      new Foo {
+        a: new Bar {
+          x: cat,
+          y: dog
+        },
+        b: foo,
+        c: 5
+      }
+      """.trimIndent())
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "a" to Deduction.Scope(
+          "x" to Deduction.Equal("cat"),
+          "y" to Deduction.Equal("dog")
+        ),
+        "b" to Deduction.Equal("foo"),
+        "c" to Deduction.EMPTY
+      )
+    )
+  }
+
+  @Test
+  fun new_deeply_nested() {
+    val expr = PaxelParser.parse(
+      """
+      new Foo {
+        x: new Bar {
+          y: new Baz {
+            z: new Buz {
+              a: foo
+            }
+          }
+        }
+      }
+      """.trimIndent()
+    )
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "x" to Deduction.Scope(
+          "y" to Deduction.Scope(
+            "z" to Deduction.Scope(
+              "a" to Deduction.Equal("foo")
+            )
+          )
+        )
+      )
+    )
+  }
+
+  @Test
+  fun from_select() {
+    val expr = PaxelParser.parse("from f in foo select f")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Equal("foo"))
+  }
+
+  @Test
+  fun from_select_field() {
+    val expr = PaxelParser.parse("from f in foo.input.baz select f.x.bar")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(Deduction.Equal("foo", "input", "baz", "x", "bar"))
+  }
+
+  @Test
+  fun from_select_binop() {
+    val expr = PaxelParser.parse("from f in foo select f.x + f.y")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Derive(
+        listOf("foo", "x"),
+        listOf("foo", "y")
+      )
+    )
+  }
+
+  @Test
+  fun from_select_new() {
+    val expr = PaxelParser.parse("from f in foo select new Bar {x: f.x}")
+
+    val actual = expr.accept(ExpressionClaimDeducer(), Unit)
+
+    assertThat(actual).isEqualTo(
+      Deduction.Scope(
+        "x" to Deduction.Equal("foo", "x")
+      )
+    )
+  }
+
+  @Test
+  fun nested_from_select() {
```

Comment:
The reason why we introduced this test is to check that `y` and `z` in the path don't get substituted with their aliases. It would be good to convey that somehow in the test name or in a comment, otherwise someone in the future could rewrite the test with different variable names because they thought this is just testing "nested_from_select", which is not the entire truth.